### PR TITLE
Allow browsing skipped logfile content with chunk navigation

### PR DESCRIPTION
**Pull request by vetri15**
This PR implements the [#4342](https://github.com/codecentric/spring-boot-admin/issues/4342) enhancement , 

other minor tweaks: 
- additionally The Buttons near the Follow mode toggle is changed from Plain SVG to font icon.
- added i18n language support text for some missing text

### CODE REVIEW

**Comment by cdprete** [comment]
@SteKoe can we get this reviewed please?

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue [hunk 1]
```diff
@@ -48,37 +48,54 @@
...
-                viewBox="0 0 24 24"
-                xmlns="http://www.w3.org/2000/svg"
-              >
-                <path
-                  d="M7 11l5-5m0 0l5 5m-5-5v12"
-                  stroke-linecap="round"
-                  stroke-linejoin="round"
-                />
-              </svg>
+            <sba-button
```
Comment (cdprete) [inline spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:51]: Maybe the `SbaIconButton` could be reused for these 3 buttons.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue [hunk 2]
```diff
@@ -136,34 +170,60 @@ export default {
...
   computed: {
     skippedBytesString() {
-      if (this.skippedBytes != null) {
+      if (this.skippedBytes != null && this.skippedBytes > 0) {
         return `skipped ${prettyBytes(this.skippedBytes)}`;
       }
       return '';
     },
+    isFollowing() {
+      return this.mode === 'follow';
```
Comment (cdprete) [inline spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Probably `mode` should be an enum.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue [hunk 3]
```diff
@@ -136,34 +170,60 @@ export default {
...
+    totalBytes: 0,
+    isChunkLoading: false,
+    faArrowDown,
+    faArrowsDownToLine,
+    faArrowUp,
   }),
   computed: {
     skippedBytesString() {
-      if (this.skippedBytes != null) {
+      if (this.skippedBytes != null && this.skippedBytes > 0) {
```
Comment (cdprete) [inline spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: I would still show it as before

**Comment by cdprete** [comment]
Hi @vetri15.

I tested this a bit, by changing the servlet sample logging configuration to
```yaml
logging:
  level:
    root: trace # even debug is fine
```
 and I found the following bugs:

- clicking the up arrow button twice indeed disables the following mode, which is in contrast the previous implementation. 
- if you try to navigate to the top, by continuously pressing the up button, it then starts to fail sometimes with
image

- lines are broken
image
This would be difficult to fix. The only way I can think about is to always fetch:

-- the previous chunk and throw away everything that comes before the last new line. If it's the first chunk, nothing will be thrown away.
-- the next chunk and throw away everything that comes after the first new line. If it's the last chunk, nothing will be thrown away.
-- handle in a special way the first line of this new fetched chunk to prepend, if it's the previous chunk, respectively append, if it's the next chunk, it to the first, if it's the previous chunk, respectively last, if it's the next chunk, line of the current chunk.

- going to the very top sometimes makes the view shrink and not always keep the full height (maybe it was a bug already existing but never visible since it was not possible to get chunks back)
image

- if the follow mode is disabled and, later on, it gets reactivated, then the number of skipped bytes is wrong
image
image

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue [hunk 4]
```diff
@@ -176,33 +236,62 @@ export default {
...
-            lines.forEach((line) => {
-              const row = document.createElement('tr');
-              const col = document.createElement('td');
-              const pre = document.createElement('pre');
-              pre.innerHTML = autolink(this.ansiUp.ansi_to_html(line));
-              col.appendChild(pre);
-              row.appendChild(col);
-              document.querySelector('.log-viewer > table')?.appendChild(row);
-            });
+            this.renderedLines = [...this.renderedLines, ...lines];
```
Comment (cdprete) [inline spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Does this imply we're keeping potentially the full log content in memory?
If it's so, that's not ideal at all.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue [hunk 5]
```diff
@@ -176,33 +236,62 @@ export default {
...
+
+            const shouldKeepAtBottom = shouldScrollToBottom || this.atBottom;
+            shouldScrollToBottom = false;
 
-            if (this.atBottom) {
-              this.scrollToBottom();
+            if (shouldKeepAtBottom) {
+              this.$nextTick(() => this.scrollToBottom());
```
Comment (cdprete) [inline spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:299]: Why do we need to use `$nextTick` now?

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 1 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:51]: I tried using **sba-icon-button**, but it does not match the required button styling because the component enforces a transparent background. Based on the current implementation, this appears to be intentional behavior of sba-icon-button. Also, in other places, similar icons are wrapped inside **sba-button**, so I kept the same approach here for consistency.
```
<template>
  <sba-button
    class="border-none sm:m-1 bg-transparent!"
    :title="title"
    size="xs"
    v-on="$attrs"
  >
    <font-awesome-icon :icon="icon" :size="size" :class="iconClass" />
  </sba-button>
</template>
```
so we have to stick with **sba-button**

----
**Suggestion.**

If this button markup feels too large, I can reduce it slightly by moving the label-selection logic into a helper function. Since the logic is only used in two places, I kept it inline for now. 
If you prefer the helper-function approach, please let me know ,I can update it accordingly.

```
isFollowing ? 'instances.logfile.page_down' : 'instances.logfile.next_chunk'
```

```
<sba-button
              :aria-label="
                $t(
                  isFollowing
                    ? 'instances.logfile.page_down'
                    : 'instances.logfile.next_chunk',
                )
              "
              :disabled="!canPageDown"
              :title="
                $t(
                  isFollowing
                    ? 'instances.logfile.page_down'
                    : 'instances.logfile.next_chunk',
                )
              "
              @click="pageDown"
            >
              <font-awesome-icon :icon="faArrowDown" />
</sba-button>
```

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:51]: No, it's good. 
It was just a suggestion.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 2 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Okay , I have changed them to Enum and assigned accordingly
```
const LogfileMode = Object.freeze({
  FOLLOW: 'follow',
  MANUAL: 'manual',
});

```

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 3 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Sure , I have reverted back to showing skipped bytes at all times.

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Please make sure to push your changes. I see no difference than before.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 2 above)
Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Please make sure to push your changes. I see no difference than before.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 4 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Hi @cdprete 

I had the same thought. However, in the previous implementation, when navigating to the logfile view, it initially fetched around 300 KB, and while staying on the page it continued accumulating newly received log lines. The loaded content was only reset when the page was refreshed or when the user left the logfile view and opened it again.

Should I change this behavior so the logfile view always maintains a 300 KB sliding window instead?

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > Should I change this behavior so the logfile view always maintains a 300 KB sliding window instead?

Mmm I don't know.
@SteKoe @ulischulte your thoughts?

I want to avoid that the memory explodes with big log files.

> However, in the previous implementation, when navigating to the logfile view, it initially fetched around 300 KB, and while staying on the page it continued accumulating newly received log lines.

But, if I'm not mistaken, before they were kept only in the dom as table rows and columns, while now we're also storming them as data.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 3 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: My bad, @cdprete. I thought I was supposed to push once all comments were resolved.

I have pushed the latest changes now, so they should be visible.

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: No problem @vetri15.

I personally prefer to have a comment in a PR closed only once the commit fixing it is pushed, otherwise it may be missed.
But that's just my personal preference.

Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue]: Understood, @cdprete. Will keep that in mind.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 4 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Hi @cdprete 

Yes, that makes sense. Keeping the log lines both in Vue state and in the rendered DOM does mean the content is effectively retained in two places.

However, directly appending rows to the DOM and then clearing the Vue state would make the DOM diverge from Vue’s source of truth, which seems risky and harder to maintain. Also, the expensive part is usually rendering and managing a large number of DOM nodes, so clearing only the JavaScript array would not fully solve the memory/performance concern.

I think the cleaner approach is to keep the rendered content itself bounded to 300kb sliding window limit.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 5 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:299]: We need that `$nextTick` because the log rows are now rendered from `renderedLines` via Vue instead of being appended to the DOM manually. Updating `renderedLines` queues the DOM patch, so calling `scrollToBottom()` synchronously would read the previous `scrollHeight`. Waiting for the next tick ensures the new `<tr>` rows are in the DOM before we scroll.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 4 above)
Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Having such an hard limit would be risky, because it highly depends on how many lines can be rendered.
On large or high-density screens, it may not be enough.

Maybe the best would be to use something like https://github.com/hilongjw/vue-recyclerview or https://github.com/Akryum/vue-virtual-scroller (which is highly used in Android, for example, https://developer.android.com/develop/ui/views/layout/recyclerview) to reuse the existing DOM nodes rather than always creating and adding new ones.
This, together with proactively dropping a batch of non visible lines from the component state when a new chunk is fetched should keep the memory quite under control even with very big log files.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 5 above)
Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:299]: Ok, understood. Thanks.

File: spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue (hunk 4 above)
Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Even if we virtualize the DOM, keeping the logfile unbounded in Vue state is still not ideal. though the data is not rendered in DOM , they still take space in Javascript. 
I think we should cap the in-memory buffer, for example around 2 MB or a fixed line count on JS variable , but only render around 300kb ~ 2000 lines (depends on the type of logs) and use range requests for older chunks.

would this be ideal ?

Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > [https://github.com/codecentric/spring-boot-admin/pull/5326#issuecomment-4366083322](https://github.com/codecentric/spring-boot-admin/pull/5326#issuecomment-4366083322)

Regarding the the "data fetch failing" error , i was able to replicate it 
1. set a small logfile size
```
  logback:
    rollingpolicy:
      max-file-size: 10MB
```
2. and open the application

now when you navigate back after the old log file is compressed and a new file takes it place , the old bytes start and end values would be irrelevant as all logs are now written in a new file and only it is exposed.

Solution:
when we get invalid byte range requested error , we can reset the logview table and track the new file properly,
cons: 
user will have no control on when the file is reset , suppose they are actively looking at some log lines , it would be refreshed without notice.

@cdprete , shall i implement this kind of solution


---
regarding the second part of that issue: "if the follow mode is disabled and, later on, it gets reactivated, then the number of skipped bytes is wrong"

I will look into this bug too , if suppose it persists after correcting the logfile range issue.

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > Even if we virtualize the DOM, keeping the logfile unbounded in Vue state is still not ideal. though the data is not rendered in DOM , they still take space in Javascript. I think we should cap the in-memory buffer, for example around 2 MB or a fixed line count on JS variable , but only render around 300kb ~ 2000 lines (depends on the type of logs) and use range requests for older chunks.
> 
> would this be ideal ?

> keeping the logfile unbounded in Vue state is still not ideal.

That's why I said: [...] proactively dropping a batch of non visible lines from the component state [...]

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > shall i implement this kind of solution

I can't say, because it would be a breaking change.
@SteKoe @ulischulte?

I think that, if you can show a notification to the user indicating that the content has been reloaded because the log file has been rotated, it's then safe enough.

Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: okay thanks @cdprete , I understand now.

1. I will implement such that we are rendering only a small portion on the screen (~300kb) and a higher limit on the javascript (~2MB) so scrolling few chunks would be faster will leverage the vue libraries you provided, both will be constants so we can set them for what's apt in prod.

2. I will show a notification after reloading logview table.

Comment (vetri15) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: Hi @cdprete 

Regarding the table shrinking issue you mentioned, the current implementation always renders the exact chunk returned by the requested range, without overlapping it with the previously rendered chunk.

For example, if the current view shows lines 1–100 and the next chunk only contains lines 101–120 because we are near the start or end of the file, only those available lines are rendered.

I kept it this way so that each previous/next chunk action consistently moves to a new byte range without mixing content from the previous chunk. If we append or prepend part of the old chunk to keep the table height stable, the navigation would no longer represent a clean chunk progression.

So the trade-off is between:
- preserving consistent chunk progression
- preserving consistent visual table height

The current implementation favors consistent chunk progression. 

if you prefer table consistency , please let me know , I will update the code

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > will leverage the vue libraries you provided, both will be constants so we can set them for what's apt in prod

They're not mine :)
Moreover, the recycle view pattern/idea is used a lot in Android, but I don't know if there are better options in Vue.
I let you and the team decide on it.

Comment (cdprete) [reply spring-boot-admin-server-ui/src/main/frontend/views/instances/logfile/index.vue:293]: > if you prefer table consistency , please let me know , I will update the code

If also the existing implementation presents the same behavior, then so be it.

**Comment by cdprete** [comment]
@vetri15 by the way, let me know if I should test it again ;)

**Comment by vetri15** [comment]
> @vetri15 by the way, let me know if I should test it again ;)

Thanks, @cdprete. I’m still working on a few edge cases, so it is taking a bit longer than expected. I’ll let you know once it’s ready for test.

**Comment by cdprete** [comment]
> > @vetri15 by the way, let me know if I should test it again ;)
> 
> Thanks, @cdprete. I’m still working on a few edge cases, so it is taking a bit longer than expected. I’ll let you know once it’s ready for test.

No hurry ;)

**Comment by vetri15** [comment]
Hi @cdprete,

I tried implementing virtual scrolling, but it introduces some usability issues.

- Since only visible rows are rendered, native Ctrl+F cannot search the full loaded content reliably. 
- Since the DOM rows are reused, native text selection becomes unstable. For example, if I select one row, the selection may appear to move to a different row after the next render/update.(looks buggy when you try selecting something on screen)

Fixing this properly would likely require custom search and custom selection/copy handling, which feels like too much complexity for just viewing previous logfile chunk.

I think the current approach should still be enough to support previous chunk navigation once the edge case bugs are fixed.

**Comment by cdprete** [comment]
Hi @vetri15.
Thanks for trying it out.

> Since only visible rows are rendered, native Ctrl+F cannot search the full loaded content reliably.

What do you mean exactly?
I would indeed expect the search to be executed only on the visible content.

> Since the DOM rows are reused, native text selection becomes unstable.

Indeed that's a "problem" of the recycle view concept and actions triggered outside of it.

> Fixing this properly would likely require custom search

Which is not a bad idea, especially if regex could be supported as well, but that's out of scope for now. 
@SteKoe I may open a feature request for this. 🤣

> I think the current approach should still be enough to support previous chunk navigation once the edge case bugs are fixed.

I agree.
